### PR TITLE
Transaction Signer interface

### PR DESCRIPTION
**Pull request by leoneparise**
## Description
This PR contains an implementation to sign Cosmos, Ethereum and Binance using JSON strings as input. The method will return a sign output which can contain an error, a json or encoded format.

## Testing instructions
Just run tests from wallet core.

## Types of changes

- Adds a new method to sign transactions using JSON strings.

## Checklist

- [X] Prefix PR title with `[WIP]` if necessary.
- [X] Add tests to cover changes as needed.
- [ ] Update documentation as needed.

### CODE REVIEW

File: src/Json/Signer.cpp [hunk 1]
```diff
@@ -0,0 +1,55 @@
+//
+// Created by Leone Parise on 2019-05-24.
```
Comment (hewigovens) [inline src/Json/Signer.cpp]: please follow other file headers

File: src/Json/Signer.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+//
+
+#include "Signer.h"
+#include "Base64.h"
+#include "../Cosmos/Signer.h"
+#include <google/protobuf/util/json_util.h>
+
+using namespace TW;
+using namespace TW::Json;
```
Comment (hewigovens) [inline src/Json/Signer.cpp]: can we find a better name? `Json` is too general, unlike `Bitcoin` or other chains

Comment (leoneparise) [reply src/Json/Signer.cpp]: I know but it's generic, I can sign any blockchain. Do you know a better name? I'm open to suggestions.

Comment (vikmeup) [reply src/Json/Signer.cpp]: 1) `Signer`, 
2) `TransactionSigner`
3) `ChainSigner`

Comment (hewigovens) [reply src/Json/Signer.cpp]: Any blockchains? AnySigner :)

File: tests/Json/SignerTests.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    auto output = signer.sign(TWCoinTypeBitcoin, transaction, privateKey);
+
+    ASSERT_TRUE(output.has_error());
+    ASSERT_EQ(JsonSignErrorCodeNotSupported, output.error().code());
+    ASSERT_EQ("Network not supported", output.error().description());
+}
+
+TEST(JsonSigner, InvalidJsonFormat) {
+    auto transaction = "{\"accountNumber\":\"8733\"\"chainId\":\"cosmoshub-2\"}";
+    auto privateKey = "ybCic4MZMapKX40aVw1QId2pHTMZvTgZvs2qv7e0Tjs=";
```
[the remark this answers is not in the document]
Comment (leoneparise) [reply tests/Json/SignerTests.cpp]: I'm using protobuf messages and `PrivateKey` objet now

File: src/Json/Signer.cpp (hunk 2 above)
Comment (leoneparise) [reply src/Json/Signer.cpp]: I liked AnySigner. It can sign any blockchain.

Comment (vikmeup) [reply src/Json/Signer.cpp]: I would use just `Signer` for simplicity

Comment (leoneparise) [reply src/Json/Signer.cpp]: Renamed on last commit.

File: src/proto/Signer.proto [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+package TW.Proto;
+option java_package = "wallet.core.jni.proto";
+
+message SigningInput {
+    // BIP44 network ID
+    uint32 network     = 1;
+    // JSON transaction
+    string transaction = 2;
+    // Private Key
+    bytes  private_key = 3;
```
Comment (vikmeup) [inline src/proto/Signer.proto]: I would propose to use private key as a string

File: src/proto/Signer.proto [hunk 5]
```diff
@@ -0,0 +1,31 @@
+syntax = "proto3";
+
+package TW.Proto;
+option java_package = "wallet.core.jni.proto";
+
+message SigningInput {
+    // BIP44 network ID
+    uint32 network     = 1;
```
Comment (vikmeup) [inline src/proto/Signer.proto]: Slip44 is a coin type

Comment (vikmeup) [reply src/proto/Signer.proto]: `coin_type` ?

File: src/Signer.cpp [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            }
+            break;
+        }
+        case TWCoinTypeBinance: {
+            Binance::Proto::SigningInput message;
+            parse(transaction, &message, output);
+            if (!output.has_error()) {
+                message.set_private_key(privateKey.bytes.data(), privateKey.bytes.size());
+                auto signerOutput = Binance::Signer(std::move(message)).build();
+                output.set_encoded(Base64::encode(signerOutput));
```
Comment (vikmeup) [inline src/Signer.cpp]: what's the purpose for base64 encoding? it's already encoded as hex

File: src/Signer.cpp [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+                );
+
+                auto signer = Ethereum::Signer(load(message.chain_id()));
+                signer.sign(key, ethTransaction);
+
+                auto protoOutput = Ethereum::Proto::SigningOutput();
+
+                auto encoded = Ethereum::RLP::encode(ethTransaction);
+                protoOutput.set_encoded(encoded.data(), encoded.size());
+
```
Comment (vikmeup) [inline src/Signer.cpp]: did you consider reuse `EthereumSigner?

Comment (leoneparise) [reply src/Signer.cpp]: I add a method to sign using the Proto SigningInput and return a SigningOutput.

File: src/Signer.cpp (hunk 6 above)
Comment (leoneparise) [reply src/Signer.cpp]: I didn't know how it was encoded. Now I converted as string.

File: src/proto/Signer.proto (hunk 5 above)
Comment (leoneparise) [reply src/proto/Signer.proto]: Good catch. I'll solve that.

File: src/interface/TWSigner.cpp [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+#include <TrustWalletCore/TWSigner.h>
+
+#include "Signer.h"
+
+using namespace TW;
+
+TW_Proto_SigningOutput TSignerSign(TW_Proto_SigningInput data)
```
Comment (hewigovens) [inline src/interface/TWSigner.cpp]: what's `TSignerSign`?

**Pull request merged**